File: src/main/java/com/paragon/client/systems/module/hud/HUDEditorGUI.java
package com.paragon.client.systems.module.hud;

import com.paragon.Paragon;
import com.paragon.api.util.render.RenderUtil;
import net.minecraft.client.gui.GuiScreen;
import net.minecraft.client.gui.ScaledResolution;
import org.apache.commons.lang3.ArrayUtils;

import java.awt.*;
import java.io.IOException;
import java.util.concurrent.atomic.AtomicBoolean;

public final class HUDEditorGUI extends GuiScreen {

    private boolean draggingComponent;

    @Override
    public void drawScreen(int mouseX, int mouseY, float partialTicks) {
        drawDefaultBackground();

        ScaledResolution scaledResolution = new ScaledResolution(mc);

        RenderUtil.drawRect((scaledResolution.getScaledWidth() / 2f) - 0.5f, 0, 1, scaledResolution.getScaledHeight(), new Color(255, 255, 255, 100).getRGB());
        RenderUtil.drawRect(0, (scaledResolution.getScaledHeight() / 2f) - 0.5f, scaledResolution.getScaledWidth(), 1, new Color(255, 255, 255, 100).getRGB());

        Paragon.INSTANCE.getModuleManager().getModulesThroughPredicate(module -> module instanceof HUDModule).forEach(hudModule -> {
            if (hudModule.isEnabled()) {
                ((HUDModule) hudModule).updateComponent(mouseX, mouseY);
                ((HUDModule) hudModule).render();
            }
        });

        super.drawScreen(mouseX, mouseY, partialTicks);
    }

    @Override
    protected void mouseClicked(int mouseX, int mouseY, int mouseButton) throws IOException {
        ArrayUtils.reverse(Paragon.INSTANCE.getModuleManager().getModules());

        final AtomicBoolean stopLoop = new AtomicBoolean(false); //Replace with a simple return@run when making this kt
        Paragon.INSTANCE.getModuleManager().getModulesThroughPredicate(module -> module instanceof HUDModule).forEach(hudModule -> {
            if (stopLoop.get()) {
                return;
            }
            if (hudModule.isEnabled() && !draggingComponent) {
                if (((HUDModule) hudModule).mouseClicked(mouseX, mouseY, mouseButton)) {
                    stopLoop.set(true);
                    return;
                }

                if (((HUDModule) hudModule).isDragging()) {
                    draggingComponent = true;
                }
            }
        });

        ArrayUtils.reverse(Paragon.INSTANCE.getModuleManager().getModules());

        super.mouseClicked(mouseX, mouseY, mouseButton);
    }

    @Override
    protected void mouseReleased(int mouseX, int mouseY, int state) {
        draggingComponent = false;

        Paragon.INSTANCE.getModuleManager().getModulesThroughPredicate(module -> module instanceof HUDModule).forEach(hudModule -> {
            if (hudModule.isEnabled()) {
                ((HUDModule) hudModule).mouseReleased(mouseX, mouseY, state);
            }
        });

        super.mouseReleased(mouseX, mouseY, state);
    }

    @Override
    public void onGuiClosed() {
        this.draggingComponent = false;

        Paragon.INSTANCE.getModuleManager().getModulesThroughPredicate(module -> module instanceof HUDModule).forEach(hudModule -> {
            if (hudModule.isEnabled()) {
                ((HUDModule) hudModule).mouseReleased(0, 0, 0);
            }
        });
    }
}

File: src/main/kotlin/com/paragon/client/systems/module/hud/HUDEditorGUI.kt
package com.paragon.client.systems.module.hud

import com.paragon.Paragon
import com.paragon.api.util.render.RenderUtil.drawRect
import net.minecraft.client.gui.GuiScreen
import net.minecraft.client.gui.ScaledResolution
import java.awt.Color
import java.io.IOException

/**
 * @author SooStrator1136
 */
class HUDEditorGUI : GuiScreen() {

    private var draggingComponent = false

    override fun drawScreen(mouseX: Int, mouseY: Int, partialTicks: Float) {
        drawDefaultBackground()

        val scaledResolution = ScaledResolution(mc)

        drawRect(
            scaledResolution.scaledWidth / 2f - 0.5f,
            0f,
            1f,
            scaledResolution.scaledHeight.toFloat(),
            Color(255, 255, 255, 100).rgb
        )
        drawRect(
            0f,
            scaledResolution.scaledHeight / 2f - 0.5f,
            scaledResolution.scaledWidth.toFloat(),
            1f,
            Color(255, 255, 255, 100).rgb
        )

        Paragon.INSTANCE.moduleManager.getModulesThroughPredicate { it is HUDModule && it.isEnabled }.forEach {
            (it as HUDModule).updateComponent(mouseX, mouseY)
            it.render()
        }

        super.drawScreen(mouseX, mouseY, partialTicks)
    }

    @Throws(IOException::class)
    override fun mouseClicked(mouseX: Int, mouseY: Int, mouseButton: Int) {
        Paragon.INSTANCE.moduleManager.modules.reverse()

        run {
            Paragon.INSTANCE.moduleManager.getModulesThroughPredicate { it is HUDModule && it.isEnabled }.forEach {
                if (!draggingComponent) {
                    if ((it as HUDModule).mouseClicked(mouseX, mouseY, mouseButton)) {
                        return@run
                    }

                    if (it.isDragging) {
                        draggingComponent = true
                    }
                }
            }
        }

        Paragon.INSTANCE.moduleManager.modules.reverse()

        super.mouseClicked(mouseX, mouseY, mouseButton)
    }

    override fun mouseReleased(mouseX: Int, mouseY: Int, state: Int) {
        draggingComponent = false
        Paragon.INSTANCE.moduleManager.getModulesThroughPredicate { it is HUDModule && it.isEnabled }.forEach {
            (it as HUDModule).mouseReleased(mouseX, mouseY, state)
        }

        super.mouseReleased(mouseX, mouseY, state)
    }

    override fun onGuiClosed() {
        draggingComponent = false
        Paragon.INSTANCE.moduleManager.getModulesThroughPredicate { it is HUDModule && it.isEnabled }.forEach {
            (it as HUDModule).mouseReleased(0, 0, 0)
        }
    }

}